src/board/v3/nap/track_channel.c

```diff
@@ -0,0 +1,152 @@
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include "board.h"
+#include "nap/nap_common.h"
+#include "nap/track_channel.h"
+#include "track.h"
+
+#include <ch.h>
+
+#include <libswiftnav/constants.h>
+#include <libswiftnav/common.h>
+#include <libswiftnav/signal.h>
+#include <libswiftnav/track.h>
+
+#include <assert.h>
+#include <string.h>
+
+#define TRACK_SAMPLE_FREQ (SAMPLE_FREQ / 4)
+
+/* NAP track channel parameters. */
+#define NAP_TRACK_CARRIER_FREQ_WIDTH              32
+#define NAP_TRACK_CODE_PHASE_FRACTIONAL_WIDTH     32
+
+#define NAP_TRACK_CARRIER_FREQ_UNITS_PER_HZ       \
+  (((u64)1 << NAP_TRACK_CARRIER_FREQ_WIDTH) / (double)TRACK_SAMPLE_FREQ)
+
+#define NAP_TRACK_CODE_PHASE_UNITS_PER_CHIP       \
+  ((u64)1 << NAP_TRACK_CODE_PHASE_FRACTIONAL_WIDTH)
+
+#define NAP_TRACK_CODE_PHASE_RATE_UNITS_PER_HZ    \
+  (NAP_TRACK_CODE_PHASE_UNITS_PER_CHIP / TRACK_SAMPLE_FREQ)
+
+#define SPACING_HALF_CHIP         ((u16)(TRACK_SAMPLE_FREQ / GPS_CA_CHIPPING_RATE) / 2)
+
+BSEMAPHORE_DECL(timing_strobe_sem, TRUE);
+
+static struct {
+  u32 code_phase;
+  u32 len;
+} nap_ch_state[NAP_MAX_N_TRACK_CHANNELS];
+
+static u32 calc_length_samples(u8 codes, s32 cp_start, u32 cp_rate)
+{
+  u16 chips = codes * 1023;
+  u64 cp_units = (1ULL << NAP_TRACK_CODE_PHASE_FRACTIONAL_WIDTH) * chips
+                 - cp_start;
+  u32 samples = cp_units / cp_rate;
+  return samples;
+}
+
+void nap_track_init(u8 channel, gnss_signal_t sid, u32 ref_timing_count,
+                   float carrier_freq, float code_phase)
+{
+  u32 now = NAP->TIMING_COUNT;
+  u32 track_count = now + 200000;
+  double cp = propagate_code_phase(code_phase, carrier_freq,
+                                  track_count - ref_timing_count);
+
+  /* Contrive for the timing strobe to occur at or close to a
+   * PRN edge (code phase = 0) */
+  track_count += (SAMPLE_FREQ/GPS_CA_CHIPPING_RATE) * (1023.0-cp) *
+                 (1.0 + carrier_freq / GPS_L1_HZ);
+
+  u8 prn = sid.sat - 1;
+  NAP->TRK_CH[channel].CONTROL = prn << NAP_TRK_CONTROL_PRN_Pos;
+  /* We always start at zero code phase */
+  NAP->TRK_CH[channel].CODE_INIT_INT = 0;
+  NAP->TRK_CH[channel].CODE_INIT_FRAC = 0;
+  NAP->TRK_CH[channel].CODE_INIT_G1 = 0x3ff;
+  NAP->TRK_CH[channel].CODE_INIT_G2 = 0x3ff;
+
+  NAP->TRK_CH[channel].SPACING = (SPACING_HALF_CHIP << 16) | SPACING_HALF_CHIP;
+
+  u32 cp_rate = (1.0 + carrier_freq/GPS_L1_HZ) * GPS_CA_CHIPPING_RATE *
+                NAP_TRACK_CODE_PHASE_RATE_UNITS_PER_HZ;
+
+  NAP->TRK_CH[channel].CARR_PINC = -carrier_freq *
+                                   NAP_TRACK_CARRIER_FREQ_UNITS_PER_HZ;
+  NAP->TRK_CH[channel].CODE_PINC = cp_rate;
+  nap_ch_state[channel].len = NAP->TRK_CH[channel].LENGTH = calc_length_samples(1, 0, cp_rate)+1;
+  nap_ch_state[channel].code_phase = (NAP->TRK_CH[channel].LENGTH) * cp_rate;
+  NAP->TRK_CONTROL |= (1 << channel); /* Set to start on the timing strobe */
```

Comment:
Add a compiler barrier before enabling the channel to make sure that all the `nap_ch_state` writes are done prior.